tensorboard/webapp/widgets/line_chart_v2/line_chart_component.ts

```diff
@@ -0,0 +1,348 @@
+
+  domDimensions: DomDimensions = {
+    main: {width: 0, height: 0},
+    xAxis: {width: 0, height: 0},
+    yAxis: {width: 0, height: 0},
+  };
+
+  private lineChart?: Chart;
+  private dataExtent: Extent = DEFAULT_EXTENT;
+  private isDataUpdated = false;
+  private isMetadataUpdated = false;
+  // Must set the default view extent since it is an optional input.
+  private isViewBoxUpdated = true;
+  private isFixedViewBoxUpdataed = false;
+  private maybeSetViewBoxToDefault = true;
+  private scaleUpdated = false;
+
+  constructor(private readonly changeDetector: ChangeDetectorRef) {}
+
+  ngOnChanges(changes: SimpleChanges) {
+    // OnChanges only decides whether props need to be updated and do not directly update
+    // the line chart.
+
+    if (changes['xScaleType']) {
+      this.xScale = createScale(this.xScaleType);
+      this.scaleUpdated = true;
+    }
+
+    if (changes['yScaleType']) {
+      this.yScale = createScale(this.yScaleType);
+      this.scaleUpdated = true;
+    }
+
+    if (changes['seriesData']) {
+      this.isDataUpdated = true;
+    }
+
+    if (changes['defaultViewExtent']) {
+      this.isFixedViewBoxUpdataed = true;
+    }
+
+    if (changes['seriesMetadataMap']) {
+      this.isMetadataUpdated = true;
+    }
+
+    this.maybeSetViewBoxToDefault = this.shouldResetViewExtent(changes);
+
+    this.updateLineChart();
+  }
+
+  ngAfterViewInit() {
+    this.initializeChart();
+    this.updateLineChart();
+  }
+
+  onViewResize() {
+    if (!this.lineChart) return;
+
+    this.readAndUpdateDomDimensions();
+    this.lineChart.resize(this.domDimensions.main);
+    this.changeDetector.detectChanges();
+  }
+
+  private shouldResetViewExtent(changes: SimpleChanges): boolean {
+    if (changes['xScaleType'] || changes['yScaleType']) {
+      return true;
+    }
+
+    const prevDefaultExtent = this.getDefaultViewBox();
+    const wasViewExtentChanged = !areExtentsEqual(
+      prevDefaultExtent,
+      this.viewBox
+    );
+
+    // Don't modify view extent if user has manually changed the view box.
+    if (wasViewExtentChanged) {
+      return false;
+    }
+
+    if (changes['seriesData']) {
+      return true;
+    }
+
+    const seriesMetadataChange = changes['seriesMetadataMap'];
+    if (seriesMetadataChange) {
+      const prevMetadataMap = seriesMetadataChange.previousValue;
+      for (const [id, metadata] of Object.entries(this.seriesMetadataMap)) {
+        const prevMetadata = prevMetadataMap && prevMetadataMap[id];
+        if (!prevMetadata || metadata.visible !== prevMetadata.visible) {
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  private initializeChart() {
+    if (this.lineChart) {
+      throw new Error('LineChart should not be initialized multiple times.');
+    }
+
+    const rendererType = this.getRendererType();
+    // Do not yet need to subscribe to the `onDrawEnd`.
+    const callbacks: ChartCallbacks = {onDrawEnd: () => {}};
```

Comment:
It was mainly for instrumentation purposes. It is currently unused and it may be removed eventually. However, it may be a good idea to instrument and send timing analytics event to track regressions.